lib/modules/onc_program/onc_standalone_public_launch_sequence.rb

```diff
@@ -0,0 +1,188 @@
+        request authorization for the provided scope from the authorization
+        endpoint, ultimately receiving an authorization token which can be used
+        to gain access to resources on the FHIR server.
+
+        # Test Methodology
+
+        Inferno will redirect the user to the the authorization endpoint so that
+        they may provide any required credentials and authorize the application.
+        Upon successful authorization, Inferno will exchange the authorization
+        code provided for an access token.
+
+        For more information on the #{title}:
+
+        * [Standalone Launch Sequence](http://hl7.org/fhir/smart-app-launch/#standalone-launch-sequence)
+      )
+
+      def url_property
+        'onc_sl_url'
+      end
+
+      def instance_url
+        @instance.send(url_property)
+      end
+
+      def instance_client_id
+        @instance.onc_public_client_id
+      end
+
+      def instance_confidential_client
+        false
+      end
+
+      def instance_client_secret
+        ''
+      end
+
+      def instance_scopes
+        @instance.onc_public_scopes
+      end
+
+      test 'OAuth server redirects client browser to app redirect URI' do
+        metadata do
+          id '01'
+          link 'http://www.hl7.org/fhir/smart-app-launch/'
+          description %(
+            Client browser redirected from OAuth server to redirect URI of
+            client app as described in SMART authorization sequence.
+          )
+        end
+
+        @instance.save
+        @instance.update(state: SecureRandom.uuid)
+
+        oauth2_params = {
+          'response_type' => 'code',
+          'client_id' => @instance.onc_public_client_id,
+          'redirect_uri' => @instance.redirect_uris,
+          'scope' => instance_scopes,
+          'state' => @instance.state,
+          'aud' => @instance.onc_sl_url
+        }
+
+        oauth_authorize_endpoint = @instance.oauth_authorize_endpoint
+
+        assert_valid_http_uri oauth_authorize_endpoint, "OAuth2 Authorization Endpoint: \"#{oauth_authorize_endpoint}\" is not a valid URI"
+
+        oauth2_auth_query = oauth_authorize_endpoint
+
+        oauth2_auth_query += if oauth_authorize_endpoint.include? '?'
+                               '&'
+                             else
+                               '?'
+                             end
+
+        oauth2_params.each do |key, value|
+          oauth2_auth_query += "#{key}=#{CGI.escape(value)}&"
+        end
+
+        redirect oauth2_auth_query[0..-2], 'redirect'
+      end
+
+      code_and_state_received_test(index: '02')
+
+      token_endpoint_tls_test(index: '03')
+
+      invalid_code_test(index: '04')
+
+      invalid_client_id_test(index: '05')
+
+      successful_token_exchange_test(index: '06')
+
+      token_response_contents_test(index: '07')
+
+      token_response_headers_test(index: '08')
+
+      test :unauthorized_read do
```

Comment:
It's probably not the right time to do this, but this should be in the `shared_onc_launch_tests` for DRYness.